### PR TITLE
[Enhancement] FE support Join in query cache

### PR DESCRIPTION
**Pull request by satanson**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

FE support JoinNode in query cache

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 33 / 66 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/EmptySetNode.java|0|2|00.00%|[68, 69]|
|:large_blue_circle:|com/starrocks/planner/FragmentNormalizer.java|15|34|44.12%|[252, 261, 262, 273, 275, 502, 509, 564, 565, 566, 567, 568, 569, 571, 572, 574, 576, 577, 655]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|17|29|58.62%|[782, 783, 847, 848, 849, 850, 851, 852, 853, 854, 855, 856]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 247 / 313 (78.91%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/AnalyticEvalNode.java|0|3|00.00%|[311, 312, 313]|
|:large_blue_circle:|com/starrocks/planner/AggregationNode.java|0|4|00.00%|[356, 357, 358, 359]|
|:large_blue_circle:|com/starrocks/planner/SortNode.java|0|3|00.00%|[289, 290, 292]|
|:large_blue_circle:|com/starrocks/planner/SetOperationNode.java|0|10|00.00%|[369, 370, 371, 372, 373, 374, 375, 376, 378, 379]|
|:large_blue_circle:|com/starrocks/planner/stream/StreamAggNode.java|0|4|00.00%|[115, 116, 117, 118]|
|:large_blue_circle:|com/starrocks/planner/EmptySetNode.java|0|2|00.00%|[81, 82]|
|:large_blue_circle:|com/starrocks/planner/HashJoinNode.java|10|13|76.92%|[149, 153, 164]|
|:large_blue_circle:|com/starrocks/planner/FragmentNormalizer.java|152|188|80.85%|[149, 150, 153, 154, 237, 244, 245, 246, 247, 249, 250, 345, 346, 357, 359, 591, 653, 654, 655, 656, 657, 658, 660, 661, 663, 665, 666, 744, 783, 849, 877, 889, 890, 891, 898, 902]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|51|52|98.08%|[758]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalOlapScanOperator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/SlotRef.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/AddDecodeNodeForDictStringRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/OlapScanImplementationRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/PlanNode.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PartitionPredicatePrune.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/ProjectNode.java|5|5|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 248 / 313 (79.23%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/AnalyticEvalNode.java|0|3|00.00%|[311, 312, 313]|
|:large_blue_circle:|com/starrocks/planner/AggregationNode.java|0|4|00.00%|[356, 357, 358, 359]|
|:large_blue_circle:|com/starrocks/planner/SortNode.java|0|3|00.00%|[289, 290, 292]|
|:large_blue_circle:|com/starrocks/planner/SetOperationNode.java|0|10|00.00%|[369, 370, 371, 372, 373, 374, 375, 376, 378, 379]|
|:large_blue_circle:|com/starrocks/planner/stream/StreamAggNode.java|0|4|00.00%|[115, 116, 117, 118]|
|:large_blue_circle:|com/starrocks/planner/EmptySetNode.java|0|2|00.00%|[81, 82]|
|:large_blue_circle:|com/starrocks/planner/HashJoinNode.java|10|13|76.92%|[149, 153, 164]|
|:large_blue_circle:|com/starrocks/planner/FragmentNormalizer.java|153|188|81.38%|[150, 151, 154, 155, 238, 245, 246, 247, 248, 250, 251, 349, 350, 361, 363, 595, 657, 658, 659, 660, 661, 662, 664, 665, 667, 669, 670, 747, 787, 853, 881, 893, 894, 895, 906]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|51|52|98.08%|[777]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalOlapScanOperator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/SlotRef.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/AddDecodeNodeForDictStringRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/OlapScanImplementationRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/PlanNode.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PartitionPredicatePrune.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/ProjectNode.java|5|5|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 303 / 350 (86.57%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/AnalyticEvalNode.java|0|3|00.00%|[311, 312, 313]|
|:large_blue_circle:|com/starrocks/planner/stream/StreamAggNode.java|0|4|00.00%|[115, 116, 117, 118]|
|:large_blue_circle:|com/starrocks/planner/EmptySetNode.java|0|2|00.00%|[81, 82]|
|:large_blue_circle:|com/starrocks/planner/SortNode.java|2|3|66.67%|[290]|
|:large_blue_circle:|com/starrocks/planner/HashJoinNode.java|10|13|76.92%|[149, 153, 164]|
|:large_blue_circle:|com/starrocks/planner/FragmentNormalizer.java|177|209|84.69%|[151, 152, 155, 156, 190, 248, 255, 256, 257, 258, 260, 261, 359, 360, 371, 373, 605, 667, 668, 669, 670, 671, 672, 674, 675, 677, 679, 680, 701, 778, 820, 939]|
|:large_blue_circle:|com/starrocks/planner/AggregationNode.java|8|9|88.89%|[356]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|51|52|98.08%|[777]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalOlapScanOperator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/RepeatNode.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/SlotRef.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/SetOperationNode.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/AddDecodeNodeForDictStringRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/OlapScanImplementationRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/PlanNode.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PartitionPredicatePrune.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/ProjectNode.java|5|5|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 298 / 346 (86.13%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/AnalyticEvalNode.java|0|3|00.00%|[312, 313, 314]|
|:large_blue_circle:|com/starrocks/planner/stream/StreamAggNode.java|0|4|00.00%|[115, 116, 117, 118]|
|:large_blue_circle:|com/starrocks/planner/EmptySetNode.java|0|2|00.00%|[81, 82]|
|:large_blue_circle:|com/starrocks/planner/SortNode.java|1|2|50.00%|[292]|
|:large_blue_circle:|com/starrocks/planner/AggregationNode.java|6|8|75.00%|[366, 368]|
|:large_blue_circle:|com/starrocks/planner/HashJoinNode.java|10|13|76.92%|[149, 153, 164]|
|:large_blue_circle:|com/starrocks/planner/FragmentNormalizer.java|177|209|84.69%|[151, 152, 155, 156, 190, 248, 255, 256, 257, 258, 260, 261, 359, 360, 371, 373, 605, 667, 668, 669, 670, 671, 672, 674, 675, 677, 679, 680, 701, 778, 820, 939]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|51|52|98.08%|[777]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalOlapScanOperator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/RepeatNode.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/SlotRef.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/SetOperationNode.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/AddDecodeNodeForDictStringRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/OlapScanImplementationRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/PlanNode.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PartitionPredicatePrune.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/ProjectNode.java|5|5|100.00%|[]|

File: gensrc/thrift/Planner.thrift [hunk 1]
```diff
@@ -52,6 +52,7 @@ struct TCacheParam {
 7: optional i64 entry_max_rows;
 8: optional bool can_use_multiversion;
 10:optional Types.TKeysType keys_type;
+11: optional set<i32> participants;
```
[the remark this answers is not in the document]
Comment (satanson) [reply gensrc/thrift/Planner.thrift:55]: comment

File: fe/fe-core/src/test/java/com/starrocks/planner/QueryCacheTest.java [hunk 2]
```diff
@@ -429,6 +462,16 @@ private void testHelper(List<String> queryList) {
                         .collect(Collectors.toList());
         ByteBuffer digest = digests.get(0);
         Set<Integer> slotRemapping = slotRemappings.get(0);
+        /*
```
[the remark this answers is not in the document]
Comment (satanson) [reply fe/fe-core/src/test/java/com/starrocks/planner/QueryCacheTest.java]: ok

File: fe/fe-core/src/main/java/com/starrocks/planner/HashJoinNode.java [hunk 3]
```diff
@@ -140,4 +142,27 @@ protected void toNormalForm(TNormalPlanNode planNode, FragmentNormalizer normali
...
         normalizeConjuncts(normalizer, planNode, conjuncts);
     }
+
+    @Override
+    public void collectEquivRelation(FragmentNormalizer normalizer) {
+        if (!joinOp.isSemiJoin() && !joinOp.isInnerJoin()) {
+            return;
+        }
+        for (BinaryPredicate eq : eqJoinConjuncts) {
+            if (!eq.getOp().equals(BinaryPredicate.Operator.EQ)) {
```
[the remark this answers is not in the document]
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/planner/HashJoinNode.java:152]: here is extract equivalence relation from eqjoinConjuncts. no need to take in-predicates, <,  > into consideration.

File: be/src/exec/hash_joiner.h [hunk 4]
```diff
@@ -205,7 +207,7 @@ class HashJoiner final : public pipeline::ContextWithDependency {
 
     void _short_circuit_break() {
         if (_phase == HashJoinPhase::EOS) {
-            set_builder_finished();
+            // set_builder_finished();
```
[the remark this answers is not in the document]
Comment (satanson) [reply be/src/exec/hash_joiner.h:210]: this function set_builder_finished is useless, just keep it

File: be/src/exec/query_cache/cache_operator.cpp [hunk 5]
```diff
@@ -486,13 +487,34 @@ Status CacheOperator::reset_lane(RuntimeState* state, LaneOwnerType lane_owner)
...
+        }
+    }
+
+    if (premature_finished_idx > 0) {
+        _lane_arbiter->enable_passthrough_mode();
+        for (auto i = 0; i <= premature_finished_idx; ++i) {
+            auto& multi_op = _multilane_operators[i];
+            multi_op->set_finished(state);
+        }
+        _scan_operator->set_finished(state);
```
[the remark this answers is not in the document]
Comment (satanson) [reply be/src/exec/query_cache/cache_operator.cpp:516]: it is a short-circuit logic, commented

File: fe/fe-core/src/main/java/com/starrocks/planner/AggregationNode.java [hunk 6]
```diff
@@ -87,6 +89,8 @@ public class AggregationNode extends PlanNode {
     
     private boolean withLocalShuffle = false;
 
+    private boolean identicallyDistributed = false;
```
[the remark this answers is not in the document]
Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/planner/AggregationNode.java:92]: withLocalShuffle has other motivation, for an example, it means to turn an ExchangeNode to a LocalExchage when the cluster only contains one  BE.

Here identicallyDistributed has clear meanings: bucket shuffle join, colocate shuffle join, one-phase agg, three-phase-agg(2nd phase of four-phase agg eliminated).

File: fe/fe-core/src/test/resources/sql/ssb/customer.sql [hunk 7]
```diff
@@ -0,0 +1,19 @@
+CREATE TABLE IF NOT EXISTS `customer` (
```
[the remark this answers is not in the document]
Comment (satanson) [reply fe/fe-core/src/test/resources/sql/ssb/customer.sql:1]: put each create sql in a standalone file inside resource directory is more clear than in a constant string which contains a lot of quote mark and escape chars.

**Comment by satanson** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 303 / 349 (86.82%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/AnalyticEvalNode.java|0|3|00.00%|[312, 313, 314]|
|:large_blue_circle:|com/starrocks/planner/stream/StreamAggNode.java|0|4|00.00%|[115, 116, 117, 118]|
|:large_blue_circle:|com/starrocks/planner/EmptySetNode.java|0|2|00.00%|[81, 82]|
|:large_blue_circle:|com/starrocks/planner/SortNode.java|2|3|66.67%|[347]|
|:large_blue_circle:|com/starrocks/planner/HashJoinNode.java|10|13|76.92%|[149, 153, 164]|
|:large_blue_circle:|com/starrocks/planner/FragmentNormalizer.java|177|209|84.69%|[148, 149, 152, 153, 187, 245, 252, 253, 254, 255, 257, 258, 356, 357, 368, 370, 602, 664, 665, 666, 667, 668, 669, 671, 672, 674, 676, 677, 698, 775, 817, 936]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|51|52|98.08%|[777]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalOlapScanOperator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/AggregationNode.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/RepeatNode.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/SlotRef.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/SetOperationNode.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/AddDecodeNodeForDictStringRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/OlapScanImplementationRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/PlanNode.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PartitionPredicatePrune.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/ProjectNode.java|5|5|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/planner/PlanNode.java [hunk 8]
```diff
@@ -942,4 +949,7 @@ public List<SlotId> getOutputSlotIds(DescriptorTable descriptorTable) {
         return descriptorTable.getTupleDesc(getTupleIds().get(0)).getSlots()
                 .stream().map(SlotDescriptor::getId).collect(Collectors.toList());
     }
+
+    public void collectEquivRelation(FragmentNormalizer normalizer) {
```
Comment (stdpain) [inline fe/fe-core/src/main/java/com/starrocks/planner/PlanNode.java:953]: some comment for collectEquivRelation?

File: fe/fe-core/src/main/java/com/starrocks/planner/FragmentNormalizer.java [hunk 9]
```diff
@@ -174,10 +202,16 @@ public Integer remapSlotId(Integer slotId) {
     }
 
     public List<Integer> remapSlotIds(List<SlotId> slotIds) {
+        if (slotIds == null || slotIds.isEmpty()) {
```
Comment (stdpain) [inline fe/fe-core/src/main/java/com/starrocks/planner/FragmentNormalizer.java:205]: why slotIds may be null ?

Comment (satanson) [reply fe/fe-core/src/main/java/com/starrocks/planner/FragmentNormalizer.java:205]: cheat the compiler and make it not to panic.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 303 / 349 (86.82%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/AnalyticEvalNode.java|0|3|00.00%|[312, 313, 314]|
|:large_blue_circle:|com/starrocks/planner/stream/StreamAggNode.java|0|4|00.00%|[115, 116, 117, 118]|
|:large_blue_circle:|com/starrocks/planner/EmptySetNode.java|0|2|00.00%|[81, 82]|
|:large_blue_circle:|com/starrocks/planner/SortNode.java|2|3|66.67%|[347]|
|:large_blue_circle:|com/starrocks/planner/HashJoinNode.java|10|13|76.92%|[149, 153, 164]|
|:large_blue_circle:|com/starrocks/planner/FragmentNormalizer.java|177|209|84.69%|[148, 149, 152, 153, 187, 245, 252, 253, 254, 255, 257, 258, 356, 357, 368, 370, 602, 664, 665, 666, 667, 668, 669, 671, 672, 674, 676, 677, 698, 775, 817, 936]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|51|52|98.08%|[807]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalOlapScanOperator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/AggregationNode.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/RepeatNode.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalOlapScanOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/SlotRef.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/SetOperationNode.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/AddDecodeNodeForDictStringRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/OlapScanImplementationRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/PlanNode.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PartitionPredicatePrune.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/ProjectNode.java|5|5|100.00%|[]|

**Review by ZiheLiu — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by satanson** [comment]
run starrocks_admit_test

**Review by stdpain — APPROVED** [review]

**Pull request merged**